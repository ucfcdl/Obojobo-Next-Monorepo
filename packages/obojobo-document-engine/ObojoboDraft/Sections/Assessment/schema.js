import { Block } from 'slate'

import ParameterNode from '../../../src/scripts/oboeditor/components/parameter-node'
import SchemaViolations from '../../../src/scripts/oboeditor/util/schema-violations'

const { CHILD_TYPE_INVALID, CHILD_MIN_INVALID } = SchemaViolations

const SETTINGS_NODE = 'ObojoboDraft.Sections.Assessment.Settings'
const RUBRIC_NODE = 'ObojoboDraft.Sections.Assessment.Rubric'
const ACTIONS_NODE = 'ObojoboDraft.Sections.Assessment.ScoreActions'
const QUESTION_BANK_NODE = 'ObojoboDraft.Chunks.QuestionBank'
const PAGE_NODE = 'ObojoboDraft.Pages.Page'

const schema = {
	blocks: {
		'ObojoboDraft.Sections.Assessment': {
			nodes: [
				{ match: [{ type: SETTINGS_NODE }], min: 1 },
				{ match: [{ type: PAGE_NODE }], min: 1 },
				{ match: [{ type: QUESTION_BANK_NODE }], min: 1 },
				{ match: [{ type: ACTIONS_NODE }], min: 1 },
				{ match: [{ type: RUBRIC_NODE }] }
			],
			normalize: (editor, error) => {
				const { node, child, index } = error
				switch (error.code) {
					case CHILD_MIN_INVALID: {
						let block
						switch (index) {
							case 0:
								block = Block.create({
									type: SETTINGS_NODE
								})
								break
							case 1:
								block = Block.create({
									type: PAGE_NODE
								})
								break
							case 2:
								block = Block.create({
									type: QUESTION_BANK_NODE,
									data: { content: { choose: 1, select: 'sequential' } }
								})
								break
							case 3:
								block = Block.create({
									type: ACTIONS_NODE
								})
								break
						}
						return editor.insertNodeByKey(node.key, index, block)
					}
					case CHILD_TYPE_INVALID: {
						switch (index) {
							case 0:
								return editor.wrapBlockByKey(child.key, {
									type: SETTINGS_NODE
								})
							case 1:
								return editor.wrapBlockByKey(child.key, {
									type: PAGE_NODE
								})
							case 2:
								return editor.wrapBlockByKey(child.key, {
									type: QUESTION_BANK_NODE,
									data: { content: { choose: 1, select: 'sequential' } }
								})
							case 3:
								return editor.wrapBlockByKey(child.key, {
									type: ACTIONS_NODE
								})
						}
					}
				}
			}
		},
		'ObojoboDraft.Sections.Assessment.Settings': {
			nodes: [{ match: [{ type: 'Parameter' }], min: 2 }],
			normalize: (editor, error) => {
				const { node, child, index } = error
				switch (error.code) {
					case CHILD_MIN_INVALID: {
						if (index === 0) {
							const block = Block.create(
								ParameterNode.helpers.oboToSlate({
									name: 'attempts',
									value: 'unlimited',
									display: 'Attempts'
								})
							)
							return editor.insertNodeByKey(node.key, index, block)
						}
						const block = Block.create(
							ParameterNode.helpers.oboToSlate({
								name: 'review',
								value: 'never',
								display: 'Review',
								options: ['always', 'never', 'no-attempts-remaining']
							})
						)
						return editor.insertNodeByKey(node.key, index, block)
					}
					case CHILD_TYPE_INVALID: {
						return editor.withoutNormalizing(c => {
							c.removeNodeByKey(child.key)
							if (index === 0) {
								const block = Block.create(
									ParameterNode.helpers.oboToSlate({
										name: 'attempts',
										value: 'unlimited',
										display: 'Attempts'
									})
								)
								return c.insertNodeByKey(node.key, index, block)
							}
							const block = Block.create(
								ParameterNode.helpers.oboToSlate({
									name: 'review',
									value: 'never',
									display: 'Review',
									options: ['always', 'never', 'no-attempts-remaining']
								})
							)
							return c.insertNodeByKey(node.key, index, block)
						})
					}
				}
			}
		}
	}
}

export default schema
