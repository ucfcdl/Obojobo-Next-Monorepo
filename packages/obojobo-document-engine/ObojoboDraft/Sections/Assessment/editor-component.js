import React from 'react'
import { Block } from 'slate'

const RUBRIC_NODE = 'ObojoboDraft.Sections.Assessment.Rubric'

class Assessment extends React.Component {
	constructor(props) {
		super(props)
		this.state = this.props.node.data.get('content')
	}

	addRubric() {
		const editor = this.props.editor
		const change = editor.value.change()

		const newRubric = Block.create({
			type: RUBRIC_NODE,
			data: { content: { type: 'pass-fail' } }
		})
		change.insertNodeByKey(this.props.node.key, this.props.node.nodes.size, newRubric)

		editor.onChange(change)
	}

	render() {
		const hasRubric = this.props.node.nodes.size === 5
		return (
			<div className={'obojobo-draft--sections--assessment'}>
				{this.props.children}
				{!hasRubric ? (
					<button className={'add-rubric'} onClick={() => this.addRubric()}>
						{'Add Rubric'}
					</button>
				) : null}
			</div>
		)
	}
}

export default Assessment
