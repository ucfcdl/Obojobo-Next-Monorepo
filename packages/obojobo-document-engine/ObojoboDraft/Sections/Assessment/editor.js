import React from 'react'
import Common from 'Common'

const ASSESSMENT_NODE = 'ObojoboDraft.Sections.Assessment'
const SETTINGS_NODE = 'ObojoboDraft.Sections.Assessment.Settings'

import Node from './editor-component'
import Settings from './components/settings/editor-component'
import Converter from './converter'
import Schema from './schema'

const plugins = {
	renderNode(props, editor, next) {
		switch (props.node.type) {
			case ASSESSMENT_NODE:
				return <Node {...props} {...props.attributes} />
			case SETTINGS_NODE:
				return <Settings {...props} {...props.attributes} />
			default:
				return next()
		}
	},
	schema: Schema
}

Common.Store.registerEditorModel(ASSESSMENT_NODE, {
	name: 'Assessment',
	isInsertable: false,
	slateToObo: Converter.slateToObo,
	oboToSlate: Converter.oboToSlate,
	plugins
})

const Assessment = {
	components: {
		Node,
		Settings
	},
	helpers: Converter,
	plugins
}

export default Assessment
