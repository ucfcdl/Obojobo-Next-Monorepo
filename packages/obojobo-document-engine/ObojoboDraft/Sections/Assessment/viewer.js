import Common from 'Common'
import Viewer from 'Viewer'

const { AssessmentUtil } = Viewer.util

import adapter from './adapter'
import ViewerComponent from './viewer-component'

Common.Store.registerModel('ObojoboDraft.Sections.Assessment', {
	type: 'section',
	adapter: adapter,
	componentClass: ViewerComponent,
	selectionHandler: null,
	getNavItem(model) {
		const title = (model.title || 'Assessment').toString()

		return {
			type: 'link',
			label: title,
			contentType: 'Assessment Section',
			path: [title.toLowerCase().replace(/ /g, '-')],
			showChildren: false,
			showChildrenOnNavigation: false
		}
	},
	variables: {
		'assessment:attemptsRemaining'(textModel, viewerProps) {
			const assessmentModel = textModel.getParentOfType('ObojoboDraft.Sections.Assessment')
			if (assessmentModel.modelState.attempts === Infinity) {
				return 'unlimited'
			}

			return AssessmentUtil.getAttemptsRemaining(viewerProps.assessmentState, assessmentModel)
		},
		'assessment:attemptsAmount'(textModel) {
			const assessmentModel = textModel.getParentOfType('ObojoboDraft.Sections.Assessment')
			if (assessmentModel.modelState.attempts === Infinity) {
				return 'unlimited'
			}

			return assessmentModel.modelState.attempts
		}
	}
})
