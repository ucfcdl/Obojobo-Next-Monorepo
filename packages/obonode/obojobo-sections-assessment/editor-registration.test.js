jest.mock('obojobo-document-engine/src/scripts/common/index', () => ({
	Registry: {
		registerModel: jest.fn(),
		getItemForType: jest.fn()
	}
}))
jest.mock('./editor-component', () => global.mockReactComponent(this, 'Assessment'))
jest.mock('./schema', () => ({ mock: 'schema' }))
jest.mock('./converter', () => ({ mock: 'converter' }))
jest.mock('slate-react')

import Common from 'obojobo-document-engine/src/scripts/common'
import { Block } from 'slate'
import Assessment from './editor-registration'

const ASSESSMENT_NODE = 'ObojoboDraft.Sections.Assessment'
const PAGE_NODE = 'ObojoboDraft.Pages.Page'
const QUESTION_BANK_NODE = 'ObojoboDraft.Chunks.QuestionBank'
const ACTIONS_NODE = 'ObojoboDraft.Sections.Assessment.ScoreActions'

describe('Assessment editor', () => {
	test('plugins.renderNode renders the Assessment when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: ASSESSMENT_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Assessment.plugins.renderNode(props, null, jest.fn())).toMatchSnapshot()
	})

	test('plugins.renderNode calls next', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: 'mockNode',
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		const next = jest.fn()

		expect(Assessment.plugins.renderNode(props, null, next)).toMatchSnapshot()
		expect(next).toHaveBeenCalled()
	})

	test('getNavItem returns expected object', () => {
		const model = {
			parent: {
				children: {
					models: [{ get: () => true }]
				}
			},
			title: 'Test Title'
		}

		expect(Assessment.getNavItem(model)).toEqual({
			type: 'link',
			label: 'Test Title',
			path: ['test-title'],
			showChildren: false,
			showChildrenOnNavigation: false
		})

		model.title = null
		expect(Assessment.getNavItem(model)).toEqual({
			type: 'link',
			label: 'Assessment',
			path: ['assessment'],
			showChildren: false,
			showChildrenOnNavigation: false
		})
	})

	test('getPasteNode extracts content', () => {
		const assessment = {
			object: 'block',
			type: ASSESSMENT_NODE,
			nodes: [
				{
					object: 'block',
					type: PAGE_NODE,
					nodes: [
						{
							object: 'block',
							type: 'mock-node'
						}
					]
				},
				{
					object: 'block',
					type: QUESTION_BANK_NODE
				},
				{
					object: 'block',
					type: ACTIONS_NODE,
					nodes: [
						{
							object: 'block',
							type: 'mock-action-node',
							nodes: [
								{
									object: 'block',
									type: PAGE_NODE,
									nodes: [
										{
											object: 'block',
											type: 'mock-node'
										}
									]
								}
							]
						}
					]
				}
			]
		}
		Common.Registry.getItemForType.mockReturnValueOnce({ getPasteNode: node => node })

		expect(Assessment.getPasteNode(Block.create(assessment))).toMatchSnapshot()
	})

	test('getPasteNode extracts partial qbs', () => {
		const assessment = {
			object: 'block',
			type: ASSESSMENT_NODE,
			nodes: [
				{
					object: 'block',
					type: QUESTION_BANK_NODE
				}
			]
		}
		Common.Registry.getItemForType.mockReturnValueOnce({ getPasteNode: node => [node] })

		expect(Assessment.getPasteNode(Block.create(assessment))).toMatchSnapshot()
	})

	test('getPasteNode extracts no nodes', () => {
		const assessment = {
			object: 'block',
			type: ASSESSMENT_NODE,
			nodes: [
				{
					object: 'block',
					type: 'mock-node'
				}
			]
		}
		expect(Assessment.getPasteNode(Block.create(assessment))).toEqual([])
	})
})
