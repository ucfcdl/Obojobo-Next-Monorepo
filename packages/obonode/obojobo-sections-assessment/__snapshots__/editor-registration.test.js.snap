// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assessment editor getPasteNode extracts content 1`] = `
Array [
  Immutable.Record {
    "data": Immutable.Map {},
    "key": "2",
    "nodes": Immutable.List [],
    "type": "mock-node",
  },
  Immutable.Record {
    "data": Immutable.Map {},
    "key": "3",
    "nodes": Immutable.List [],
    "type": "ObojoboDraft.Chunks.QuestionBank",
  },
  Immutable.Record {
    "data": Immutable.Map {},
    "key": "7",
    "nodes": Immutable.List [],
    "type": "mock-node",
  },
]
`;

exports[`Assessment editor getPasteNode extracts partial qbs 1`] = `
Array [
  Immutable.Record {
    "data": Immutable.Map {},
    "key": "9",
    "nodes": Immutable.List [],
    "type": "ObojoboDraft.Chunks.QuestionBank",
  },
]
`;

exports[`Assessment editor plugins.renderNode calls next 1`] = `undefined`;

exports[`Assessment editor plugins.renderNode renders the Assessment when passed 1`] = `
<Assessment
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Sections.Assessment",
    }
  }
/>
`;
